modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/GmosControllerSim.scala

```diff
@@ -3,28 +3,49 @@
 
 package edu.gemini.seqexec.server
 
-import org.log4s._
+import java.util.concurrent.atomic.AtomicBoolean
 
+import org.log4s._
 import edu.gemini.seqexec.model.dhs.ImageFileId
 import GmosController.{GmosConfig, NorthTypes, SiteDependentTypes, SouthTypes}
 
-import scalaz.EitherT
+import scalaz.{EitherT, \/}
+import scalaz.Scalaz._
 import scalaz.concurrent.Task
 
 private class GmosControllerSim[T<:SiteDependentTypes](name: String) extends GmosController[T] {
   private val Log = getLogger
 
   override def getConfig: SeqAction[GmosConfig[T]] = ??? // scalastyle:ignore
 
+  private val stopFlag = new AtomicBoolean(false)
+  private val abortFlag = new AtomicBoolean(false)
+
+  private val tic = 200
+  private def observeTic(obsid: ImageFileId, remain: Int): Task[\/[SeqexecFailure, ImageFileId]] =
+    if(remain < tic) Task {
+      Log.info(s"Simulate Gmos $name observation completed")
+      TrySeq(obsid)
+    }
+    else Task.delay(stopFlag.get).flatMap( stop => Task.delay(abortFlag.get).flatMap( abort =>
+      if(stop) Task(TrySeq.fail(SeqexecFailure.Execution("Exposure stopped by user.")))
+      else if(abort) Task(TrySeq.fail(SeqexecFailure.Execution("Exposure aborted by user.")))
+      else Task{ Thread.sleep(tic.toLong)} *> observeTic(obsid, remain-tic)))
+
   override def observe(obsid: ImageFileId): SeqAction[ImageFileId] = EitherT( Task {
     Log.info(s"Simulate taking Gmos $name observation with label " + obsid)
-    Thread.sleep(5000)
-    Log.info(s"Simulate Gmos $name observation completed")
-    TrySeq(obsid)
-  } )
+    stopFlag.set(false)
```

Comment:
Should these actions be atomic? What happens if setting `abortFlag` fails. That would lean `stopFlag` set

Reply:
How could `set` they fail?

Reply:
is this setting an epics channel's value?

Reply:
No, `stopFlag` is an `AtomicInteger`. This is `GmosControllerSim`.

Reply:
ok, in principle they should be both be changed at once, but this is only for simulation so it should be ok